core/src/main/scala/cats/data/ContT.scala

```diff
@@ -105,6 +105,29 @@ object ContT {
       def apply[A](ma: M[A]): ContT[M, B, A] = ContT.liftF(ma)
     }
 
+  /*
+   * Call with current continuation
+   *
+   * Passes the current continuation to f, meaning we can model short-circuit
+   * evaluation eg exception handling
+   *
+   * {{{
+   *   for {
+   *     _ <- ContT.callCC( (k: Unit => ContT[IO, Unit, Unit]) =>
+   *       ContT.liftF(IO.println("this will print first")) >>
+   *         k(()) >>
+   *         ContT.liftF(IO.println("this will NOT print as we short-circuit to the contination"))
+   *     )
+   *     _ <- ContT.liftF(IO.println("this will print second")])
+   *   } yield ()
+   *
+   * }}}
+   */
+  def callCC[M[_], R, A, B](f: (A => ContT[M, R, B]) => ContT[M, R, A]): ContT[M, R, A] =
+    apply { cb =>
+      f(a => apply(_ => cb(a))).run(cb)
```

Comment:
I'm trying to see:

1. Why we can't move cb(a) out
2. If we should use Defer[M] anywhere to make this lazier or safer (as is done in tailRecM)

Reply:
Sorry, what did you mean by move `cb(a)` out?

The `Defer` thing is something I was worried about. I am not aware that we need it anywhere but was hoping that somebody with more experience would catch it in review if I'm wrong

Reply:
for instance:
```scala
apply {
  val cont = f { a =>
    val cba = cb(a)
    apply(_ => cba)
  }

  cont.run(cb)
}
```
has potentially different execution behavior, although, if we have purity, I think the same result. It isn't clear to me which is likely to be more efficient.

Secondly, if we have a `Defer[M]` then we could put the entire `cont.run(cb)` in a defer: `Defer[M].defer(cont.run(cb))`

Since we do that on map and flatMap, I assume there are recursive cases that would want that here as well, and if nothing else, for consistency, I guess I'd do it...

But it would be nice to find a recursive example and see if we can show we can do deep recursions.

Reply:
Hmm it would appear that it is not stack-safe. I pushed a failing test for it but no amount of poking it with `defer` and `AndThen` has fixed it yet 😂

Reply:
this seems to fix it:
```diff
oscar@Patricks-MacBook-Air cats % git diff
diff --git a/core/src/main/scala/cats/data/ContT.scala b/core/src/main/scala/cats/data/ContT.scala
index 4bef4e13c..d4864b80b 100644
--- a/core/src/main/scala/cats/data/ContT.scala
+++ b/core/src/main/scala/cats/data/ContT.scala
@@ -40,7 +40,10 @@ sealed abstract class ContT[M[_], A, +B] extends Serializable {
     // allocate/pattern match once
     val fnAndThen = AndThen(fn)
     ContT[M, A, C] { fn2 =>
-      val contRun: ContT[M, A, C] => M[A] = _.run(fn2)
+      val contRun: ContT[M, A, C] => M[A] =  { c =>
+        M.defer(c.run(fn2))
+      }
+
       val fn3: B => M[A] = fnAndThen.andThen(contRun)
       M.defer(run(fn3))
     }

```
in the flatMap function.

Can you try that (I pulled your PR and it worked for me).

Reply:
😍 THANK YOU! Very good call. Just out of curiosity: what made you realize that flatMap was the issue?

I've added an explicit stack-safety test for `flatMap` as well (which I just assumed already existed hence not even thinking to suspect `flatMap`)

Reply:
Honestly, I just looked at all the places a Defer could go and I tried it.

Afterwords, it seemed more obvious that flatMap needed one there: it was internally doing a run which starts a new stack on whatever M you have. By using a Defer, you should push that evaluation onto the same final heap used to interpret M.

I think a good rule of thumb should be: any time we do a `run` internally, we should wrap that inside `Defer[M].defer`.

Reply:
Thanks, just trying to visualize the execution and the interaction between `AndThen` and `Defer`. Any insights much appreciated 😂